cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -0,0 +1,739 @@
+  return {std::move(path)};
+}
+
+inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
+  if (expr.literal()) return expr;
+
+  if (auto ref = expr.field_ref()) {
+    return compute::field_ref(*ref->FindOne(*kBoringSchema));
+  }
+
+  auto modified_call = *CallNotNull(expr);
+  for (auto& arg : modified_call.arguments) {
+    arg = UseBoringRefs(arg);
+  }
+  return compute::Expression{std::move(modified_call)};
+}
+
+std::shared_ptr<Buffer> SubstraitFromJSON(util::string_view type_name,
+                                          util::string_view json) {
+  auto maybe_buf = internal::SubstraitFromJSON(type_name, json);
+  DCHECK_OK(maybe_buf.status());
+  return maybe_buf.ValueOrDie();
+}
+
+std::string SubstraitToJSON(util::string_view type_name, const Buffer& buf) {
+  auto maybe_buf = internal::SubstraitToJSON(type_name, buf);
+  DCHECK_OK(maybe_buf.status());
+  return maybe_buf.ValueOrDie();
+}
+
+TEST(Substrait, SupportedTypes) {
+  auto ExpectEq = [](util::string_view json, std::shared_ptr<DataType> expected_type) {
+    ARROW_SCOPED_TRACE(json);
+
+    ExtensionSet empty;
+    auto buf = SubstraitFromJSON("Type", json);
+    ASSERT_OK_AND_ASSIGN(auto type, DeserializeType(*buf, empty));
+
+    EXPECT_EQ(*type, *expected_type);
+
+    ASSERT_OK_AND_ASSIGN(auto serialized, SerializeType(*type, &empty));
+    EXPECT_EQ(empty.types().size(), 0);
+
+    // FIXME chokes on NULLABILITY_UNSPECIFIED
+    // EXPECT_THAT(internal::CheckMessagesEquivalent("Type", *buf, *serialized), Ok());
+
+    ASSERT_OK_AND_ASSIGN(auto roundtripped, DeserializeType(*serialized, empty));
+
+    EXPECT_EQ(*roundtripped, *expected_type);
+  };
+
+  ExpectEq(R"({"bool": {}})", boolean());
+
+  ExpectEq(R"({"i8": {}})", int8());
+  ExpectEq(R"({"i16": {}})", int16());
+  ExpectEq(R"({"i32": {}})", int32());
+  ExpectEq(R"({"i64": {}})", int64());
+
+  ExpectEq(R"({"fp32": {}})", float32());
+  ExpectEq(R"({"fp64": {}})", float64());
+
+  ExpectEq(R"({"string": {}})", utf8());
+  ExpectEq(R"({"binary": {}})", binary());
+
+  ExpectEq(R"({"timestamp": {}})", timestamp(TimeUnit::MICRO));
+  ExpectEq(R"({"date": {}})", date32());
+  ExpectEq(R"({"time": {}})", time64(TimeUnit::MICRO));
+  ExpectEq(R"({"timestamp_tz": {}})", timestamp(TimeUnit::MICRO, "UTC"));
+  ExpectEq(R"({"interval_year": {}})", interval_year());
+  ExpectEq(R"({"interval_day": {}})", interval_day());
+
+  ExpectEq(R"({"uuid": {}})", uuid());
+
+  ExpectEq(R"({"fixed_char": {"length": 32}})", fixed_char(32));
+  ExpectEq(R"({"varchar": {"length": 1024}})", varchar(1024));
+  ExpectEq(R"({"fixed_binary": {"length": 32}})", fixed_size_binary(32));
+
+  ExpectEq(R"({"decimal": {"precision": 27, "scale": 5}})", decimal128(27, 5));
+
+  ExpectEq(R"({"struct": {
+    "types": [
+      {"i64": {}},
+      {"list": {"type": {"string":{}} }}
+    ]
+  }})",
+           struct_({
+               field("", int64()),
+               field("", list(utf8())),
+           }));
+}
```

Comment:
I don't see `MAP<K, V>` utilized here.  Is it supported?  If not we should probably create a JIRA.

Reply:
Nice catch; fixed a bug with that one. c8e93ea